rfd/0226-self-healing.md

```diff
@@ -0,0 +1,286 @@
+
+For Proxy services, ideally we would distinguish between heartbeat failures and
+true agent connectivity failures to avoid unnecessary agent reconnects. However
+reliably determining whether a Proxy can still service agent connectivity is complex.
+As a pragmatic approach, we can base recovery decisions solely on the existing
+heartbeat health mechanism.
+
+### Auth Reconnects
+
+Agents create a gRPC client connection to Auth. The transport of this connection
+could be a direct TCP connection to Auth or a tunneled connection through the
+Proxy. The solution should be indifferent to the transport.
+
+We can communicate the health of the Auth instance an agent is connected to using
+gRPC health checking[^1]. This is a standardized service API that gRPC servers can implement. Clients that enable health checking can then poll or stream the health
+of the service.
+
+This solves half the problem. Agents can enable health checking and can be aware
+of when they are connected to an unhealthy Auth service. However the client will
+not automatically reconnect in this scenario. Within gRPC client connections are
+maintained by load balancing policies[^2].
+
+Evaluating the default `pick_first` policy. It will only create a new connection
+when the existing TCP connection is disrupted. When health checking is enabled
+the client begins failing requests immediately rather than sending them to the
+server but never attempts to reconnect.
+
+Evaluating the `round_robin` policy, a connection is created to each address that
+was returned by the resolver. This policy is able to send requests to healthy
+connections based on health checks. However this assumes that each address represents
+a unique server instance. This is not usually the case. Teleport is typically fronted by one or more layers of load balancers. This means there could be
+a single address or connecting to multiple addresses could result in connections
+to the same server instance.
+
+In summary, neither the `pick_first` or the `round_robin` load balancing policies
+will work for our use case. Other built-in policies are available but they also
+do not meet our use case[^3].
+
+The solution here is to implement our own load balancing policy. The go-grpc library
+conveniently allows you to register custom policies[^2] by implementing the
+balancer interface[^4]. This policy will enable gRPC health checking by default
+and use the streaming API to ensure health checks can scale to a large number of
+clients.
+
+To avoid introducing too much new behavior we will base our implementation off of
+the existing `pick_first` policy[^5]. The primary difference will be when the existing
+connection enters the `TRANSIENT_FAILURE` state[^6] new connection attempts will be
+made until either the existing connection returns to `READY` or the new connection
+reaches the `READY` state.
+
+All connection attempts should use a backoff policy with jitter to avoid a thundering
```

Comment:
Curious how we want to handle retrieving the service configuration as well. Is jitter enough to prevent issues with all agents making HTTP requests against Proxy to see if any changes to service configuration have occurred?

Also curious what Teleport agents do today if a Proxy shuts down. Not sure if a new client connection is created or how that reconnect logic works. Would we be safe to only query the find endpoint once during connect/reconnect? As of now, a Proxy server would have to restart to begin returning a different service configuration. Can we avoid polling altogether?